### ISSUE TITLE
Provide more efficient method for getting node at a specific location

### ISSUE THREAD

**Issue by zbirenbaum** [nvim-treesitter/nvim-treesitter#2765]
Recently, I was attempting to improve the efficiency of a plugin which gets unused tokens from diagnostics, then uses treesitter to get the group of those tokens, and then uses that to get the highlight and dim it.

Unfortunately, despite looking through the API for hours, I cannot find a better way of getting the information of a token at a specific location given the bufnumber, line number, and character range. The (imo absurd) code required to accomplish the task even given such detailed information, which was taken almost directly from the official repository that accomplishes a similar task is as follows:

```
dim.get_treesitter_nodes = function(buf, ns, row, col, end_col)
  local self = highlighter.active[buf]
  if not self then
    return {}
  end
  local matches = {}
  self.tree:for_each_tree(function(tstree, tree)
    if not tstree then
      return
    end
    local root = tstree:root()
    local root_start_row, _, root_end_row, _ = root:range()
    if root_start_row > row or root_end_row < row then
      return
    end
    local query = self:get_query(tree:lang())
    if not query:query() then
      return
    end
    local iter = query:query():iter_captures(root, self.bufnr, row, row + 1)
    for capture, node, _ in iter do
      local hl = query.hl_cache[capture]
      if hl and ts_utils.is_in_node_range(node, row, col) then
        local c = query._query.captures[capture]
        if c ~= nil then
          local general_hl = query:_get_hl_from_capture(capture)
          table.insert(matches, { ts_group = general_hl, node = node })
        end
      end
    end
  end, true)
  local final = #matches >= 1 and matches[#matches] or nil
  if not final then
    return
  end
  local unused_group = final.ts_group .. "Unused"
  ts_utils.highlight_node(final.node, buf, ns, unused_group)
  vim.api.nvim_buf_add_highlight(buf, ns, unused_group, row, col, end_col)
  return final.ts_group
end
```
Because of this, unlike any other treesitter highlight, ones I attempt to conditionally set in this way will continually flash wrong colors if I set update in insert for the diagnostics to on, even while using extmarks, because treesitter will return nil for the group when typing quickly.

Considering that to set the highlights at all, treesitter must have extmarks for the namespace or some way of hooking into the direct locations, providing access to that would be greatly appreciated.

**Comment by clason** [comment]
`:h nvim_buf_get_extmarks()`?

I'm sorry, but this reads like a long rant without any actionable feature request ("please make more performant").

If there's a specific function you'd like exposed, please say so.

**Comment by zbirenbaum** [comment]
> `:h nvim_buf_get_extmarks()`?
> 
> I'm sorry, but this reads like a long rant without any actionable feature request ("please make more performant" -- if we knew how, we would have already done so).
> 
> If there's a specific function you'd like exposed, please say so.

I use nvim_buf_get_extmarks() in my plugin. To my knowledge though, that cannot provide the hl group of a token or the highlight of a character within directly.

i apologize if I came across as a rant. More direct access to either the TSGroup or highlight of a token given a location is all I’m looking for.

**Comment by clason** [comment]
> i apologize if I came across as a rant. More direct access to either the TSGroup or highlight of a token given a location is all I’m looking for. Ideally in the form of creating a ‘pseudo’ TSGroup for individual nodes that will be set and unset by certain conditions.

I apologize if I'm sounding abrasive, but that's still meaningless to me -- this is just not information available to tree-sitter (as it involves how Neovim internally manages highlights).

Maybe https://github.com/nvim-treesitter/playground can be of use to you, though?

**Comment by theHamsta** [comment]
You probably don't want to connect the highlights with semantic information but rather use your own queries for the stuff you're interested in. Tree-sitter sets extmarks here with a private namespace https://github.com/theHamsta/neovim/blob/fb37dcf40f73f27c4a3b857cf1f4d86964076a1f/runtime/lua/vim/treesitter/highlighter.lua#L273 It also uses `ephemeral` extmark which disappear on edits and get set newly by re-parsing. All our code retrieving the highlights is just slow debug code. You can hook into the changes of the tree like the highlighter does to get notified about changed ranges.

**Comment by zbirenbaum** [comment]
> > i apologize if I came across as a rant. More direct access to either the TSGroup or highlight of a token given a location is all I’m looking for. Ideally in the form of creating a ‘pseudo’ TSGroup for individual nodes that will be set and unset by certain conditions.
> 
> I apologize if I'm sounding abrasive, but that's still meaningless to me -- this is just not information available to tree-sitter (as it involves how Neovim internally manages highlights).
> 
> Maybe https://github.com/nvim-treesitter/playground can be of use to you, though?

I believe I looked into the ts-playground api. The backend of it may actually be where I got this code. Should I request this in neovim then?

Essentially what I want to do is this:
Suppose a variable’s TS Group is something like TSVariable. Neovim highlights TS variables with fg=blue. 

Now, on diagnostic publish, I see the that said token is unused. I create a highlight called TSVariableUnused, which had a highlight of fg=dark blue, which treesitter interprets the same as TSVariable for parsing purposes.

now, the tree is functionally identical, but when neovim goes to highlight, until I detect that the token is no longer detected as ‘unused’ by the diagnostics, the token will never be highlighted as TSVariable, only TSVariableUnused.

If there is an alternative way to accomplish the same thing , I am very open to other ideas, but extmarks could not replicate the hypothetical behavior of what I described.

**Comment by clason** [comment]
Let me be blunt: I think what you are wishing for is just not reasonable given how things actually work.

**Comment by zbirenbaum** [comment]
> You probably don't want to connect the highlights with semantic information but rather use your own queries for the stuff you're interested in. Tree-sitter sets extmarks here with a private namespace https://github.com/theHamsta/neovim/blob/fb37dcf40f73f27c4a3b857cf1f4d86964076a1f/runtime/lua/vim/treesitter/highlighter.lua#L273 It also uses `ephemeral` extmark which disappear on edits and get set newly by re-parsing (in contrast to diagnostics that get just moved. All our code retrieving the highlights is just slow debug code. You can hook into the changes of the tree like the highlighter does to get notified about changed ranges.

I looked into this option, but creating a syntax based query for unused variables didn’t look feasible. I’ll look into this though.

**Comment by theHamsta** [comment]
You can also patch the highlight queries and add an additional predicate to some of them. This will be a Lua function that yields true/false (used/unused in your case) and will only apply the highlight if true. It might be good instead to just set your own highlights with higher priority over the built-in ones.

**Comment by zbirenbaum** [comment]
> Let me be blunt: I think what you are wishing for is just not reasonable given how things actually work.

Ah, ok. I’ll look into other options. Thank you for your time and consideration of this. I appreciate the response.

**Comment by zbirenbaum** [comment]
> You can also patch the highlight queries and add an additional predicate to some of them. This will be a Lua function that yields true/false (used/unused in your case) and will only apply the highlight if true. It might be good instead to just set your own highlights with higher priority over the built-in ones: just find unused varibales and set an own "unused" highlight.

This is a great idea! Thanks!!

**Comment by clason** [comment]
What @theHamsta mentioned sounds like the best way to achieve what you want. Good luck!

**Comment by zbirenbaum** [comment]
@theHamsta Thank you so much for the advice! The code is a bit of a mess at the moment because I was trying all sorts of stuff to test it out, but I finally managed to get the highlights to move as I typed so they look perfectly natural by using end_right_gravity, and highlighting purely through extmarks instead of any of the various highlight api functions.

Whenever diagnostics update, I clear the extmarks in the namespace, then reapply them. I kept getting annoying glitches with text on the line after the diagnostic being highlighted as 'unused' whenever I deleted a line in command mode (mostly an issue  at the bottom of long files for some reason, works at the top though, may be a bug in lsp api) so I created a timer to get the current diagnostics and update every 200ms. It now works perfectly smoothly without having update_in_insert and has none of the apparent performance cost to boot!
